packages/react-compose/src/next/README.md

```diff
@@ -0,0 +1,221 @@
+
+However, deep merge overlooks many edge cases for component props:
+
+- Deep merging classnames should append them, not replace
+- Deep merging JSX or arrays should replace, not recurse
+- Deep merging an object on a string should replace
+
+...which introduces the first utility: `mergeProps`. Merge props works like a deep merge, but takes care of classnames, JSX, arrays, and object edge cases.
+
+#### Supporting the `as` prop with `getSlots`
+
+Fluent UI components take a common `as` prop. This allows the root element to be rendered with something other than the default.
+
+To support the `as` prop, the render function might look like this:
+
+```jsx
+const renderButton = state => {
+  const root = state.as;
+
+  return <root {...state} />;
+};
+```
+
+Additionally, you will need to filter out native properties which apply to the root; otherwise you will end up mixing any unexpected props into the element. To do this, we have a `getNativeElementProps` helper, which can be used for this purpose:
+
+```jsx
+const renderButton = state => {
+  const root = state.as;
+  const rootProps = getNativeElementProps(root, state);
+
+  return <root {...rootProps} />;
+};
+```
+
+This step has been abstracted in the `getSlots` helper:
+
+```jsx
+const renderButton = state => {
+  const { slots, slotProps } = getSlots(state);
+
+  return <slots.root {...slotProps.root} />;
+};
+```
+
+#### Supporting shorthand props
+
+Fluent UI components almost always contain sub parts, and these sub parts should be configurable. We allows them to be configured through "shorthand props", which let's the caller pass in a variety of inputs for a given slot. Take a Button's "icon" slot:
+
+```jsx
+// The icon can be a string
+<Button icon="X" />
+
+// The icon can be JSX
+<Button icon={ <FooIcon/> }/>
+
+// The icon can be an object
+<Button icon={{ as: 'i', children: getCode('Add') } } />
+
+// The icon can be a children function (which receives the original slot and props)
+<Button icon={{ children: (C, p) => <C {...p} otherThings /> }} />
+```
+
+Supporting this dynamic props input requires some helpers:
+
+1. A helper `simplifyShorthand` to simplify the user's input into an object for props merging
+2. The `getSlots` helper to parse the slots out
+
+Here's how this looks:
+
+The factory function, which deep clones the props, would need to simplify the shorthand first:
+
+```jsx
+const createButton = (userProps, ref, defaultProps) => {
+  const state = mergeProps(
+    {
+      // default props
+      as: 'button',
+      ref,
+      icon: { as: 'span' },
+    },
+    defaultProps, // optional default props from the caller
+    simplifyShorthand(userProps, ['icon']), // simplify the user's props
```

Comment:
why don't we call `simplifyShorthand ` internally in `mergeProps`?  instead, pass `userProps` and `shorthandPropNames` to `mergeProps`

Reply:
I was thinking that not every component has slots, yet you'd still want to use mergeProps for merging user input on defaults.

Reply:
Also, right now mergeProps takes in ...rest as the additional sets. I am not sure how to do that without making the shorthandPropNames the first or second argument, which seems off to emphasize